### ISSUE TITLE
Discuss dissociation between deserialization and validation

### ISSUE THREAD

**Issue by raphaelrobert** [openmls/openmls#1132]
Currently, we abort deserialization when we do validation for deserialized structs. We can discuss alternatives to this flow.

**Comment by duesee** [comment]
Food for thought: [Parse, don’t validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)

Maybe all we need is a `#[validate(leaves_are_ordered)]` in tls-codec?

Spoiler:

> Consider: what is a parser? Really, a parser is just a function that consumes less-structured input and produces more-structured output. By its very nature, a parser is a partial function—some values in the domain do not correspond to any value in the range—so all parsers must have some notion of failure. Often, the input to a parser is text, but this is by no means a requirement, and parseNonEmpty is a perfectly cromulent parser: it parses lists into non-empty lists, signaling failure by terminating the program with an error message.
> 
> Under this flexible definition, parsers are an incredibly powerful tool: they allow discharging checks on input up-front, right on the boundary between a program and the outside world, and once those checks have been performed, they never need to be checked again! Haskellers are well-aware of this power, and they use many different types of parsers on a regular basis [...]

**Comment by kkohbrok** [comment]
I think we already encode guarantees in the type system in quite a few places. The validation chain is an example for that. For simple checks as in the unmerged leaves example, a macro might work. But in most cases, validation has to happen relative to some other data, so the macro might have to be a little more complex.

In #1120, @raphaelrobert alludes to another point which I think came up recently as well: These constructs are nice, but they add complexity and can create quite a bit of churn when the system changes. I'm generally in favor of having as many guarantees as possible encoded in the type system, but I think that's something that we should do once all the changes for draft-16 have landed and we know that the spec won't change to a significant degree anymore.

**Comment by duesee** [comment]
> I think we already encode guarantees in the type system in quite a few places. The validation chain is an example for that.

Oh, sure. And that's pretty great :-) With my comment above I wanted to say that maybe doing "validation tasks" during parsing is not as unclean as it feels like -- at least it always felt unclean to me. But you can define context-free grammars that are very close to what feels to be "validation". I.e., some protocols have productions in place that enforce minimum and maximum values, that one element of a certain type needs to be in a list of elements with another type¹, or that elements are ordered.

Sometimes grammars are even too restricitve, IMO. For example, you can design a production such that a base64 string always has the correct length. But this is difficult to implement and it makes more sense to just greadily consume the allowed base64 characters and call `base64_decode` as a "validation" step afterwards inside the parsing step.

Long story short: I think this is a difficult topic and we probably want to decide about how far we want messages to reach into business logic case-by-case. Maybe it makes sense to accept an invalid list to be able to send an encrypted error later? (Just a made-up example.)

¹ IMAP's `mbx-list-flags` rule for example.

**Comment by keks** [comment]
My general impression is

- It's fine to validate stuff at parse time (or immediately after, but in the type constructor) for which we don't need any context
- If we need more context, we can think about making it available to the type constructor function (e.g. a public key for verifying if it is known)
- If we need a lot of context (e.g. the entire group), do the validation checks later

And I think this is how the code looks now, right? In that case I think we can close this.
